src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
```

Comment:
You could use a mutable record here:

```ocaml
type 'a node = Nil | Next of { mutable value: 'a; mutable next: 'a node ref }
```

The advantage of this is that the size of a node is now smaller.  This can improve performance (less frequent GC needed, fewer words written to memory, GC examines fewer words).  The disadvantage is that the implemention (avoiding space leaks) is a bit a more intricate as you need to point to a node and must then clear the `value` field after reading it in `pop`.